src/ffi.c

```diff
@@ -0,0 +1,1103 @@
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+*/
+
+#include "private.h"
+
+#include <ffi.h>
+#include <stdint.h>
+
+#define C_MACRO_STRING_DEF(x) JS_PROP_STRING_DEF(#x, x, JS_PROP_CONFIGURABLE)
+
+#define JS_PTR_TYPE t_bigint
+#define JS_IS_PTR(ctx, x) JS_IsBigInt(ctx, x)
+
+#if UINTPTR_MAX == UINT32_MAX
+#define JS_TO_UINTPTR_T(ctx, pres, val) {uint64_t v; JS_ToBigInt64(ctx, &v, val); *(uint32_t *)(pres) = (uint32_t)v;}
+#define JS_NEW_UINTPTR_T(ctx, val) JS_NewBigUint64(ctx, (int32_t)(val))
+#define ffi_type_ptr ffi_type_uint32
+#elif UINTPTR_MAX == UINT64_MAX
+#define JS_TO_UINTPTR_T(ctx, pres, val) JS_ToBigInt64(ctx, (int64_t *)(pres), val)
+#define JS_NEW_UINTPTR_T(ctx, val) JS_NewBigUint64(ctx, (int64_t)(val))
+#define ffi_type_ptr ffi_type_uint64
+#else
+#error "'uintptr_t' neither 32bit nor 64 bit, I don't know how to handle it."
+#endif
+
+#define STR(x) #x
+#if SIZE_MAX == UINT32_MAX
+#define JS_TO_SIZE_T(ctx, pres, val) JS_ToInt32(ctx, (int32_t *)(pres), val)
+#define JS_NEW_SIZE_T(ctx, val) JS_NewInt32(ctx, (int32_t)(val))
+#define JS_PROP_SIZE_T_DEF(name, val) JS_PROP_INT32_DEF(name, (int32_t)(val), JS_PROP_CONFIGURABLE)
+#define C_SIZEOF_DEF(x) JS_PROP_INT32_DEF(STR(sizeof_##x), (int32_t)(sizeof(x)), JS_PROP_CONFIGURABLE)
+#define ffi_type_size_t ffi_type_uint32
+#elif SIZE_MAX == UINT64_MAX
+#define JS_TO_SIZE_T(ctx, pres, val) JS_ToInt64(ctx, (int64_t *)(pres), val)
+#define JS_NEW_SIZE_T(ctx, val) JS_NewInt64(ctx, (int64_t)(val))
+#define JS_PROP_SIZE_T_DEF(name, val) JS_PROP_INT64_DEF(name, (int64_t)(val), JS_PROP_CONFIGURABLE)
+#define C_SIZEOF_DEF(x) JS_PROP_INT64_DEF(STR(sizeof_##x), (int64_t)(sizeof(x)), JS_PROP_CONFIGURABLE)
+#define C_OFFSETOF_DEF(t, d) JS_PROP_INT64_DEF(STR(offsetof_##t##_##d), (int64_t)(offsetof(t, d)), JS_PROP_CONFIGURABLE)
+#define ffi_type_size_t ffi_type_sint32
+#else
+#error "'size_t' neither 32bit nor 64 bit, I don't know how to handle it."
+#endif
+
+#if INT_MAX == INT32_MAX
+#define C_MACRO_INT_DEF(x) JS_PROP_INT32_DEF(#x, (int32_t)(x), JS_PROP_CONFIGURABLE)
+#define C_ENUM_DEF(x) JS_PROP_INT32_DEF(#x, (int32_t)(x), JS_PROP_CONFIGURABLE)
+#define JS_TO_INT(ctx, pres, val) JS_ToInt32(ctx, (int32_t *)(pres), val)
+#define JS_NEW_INT(ctx, val) JS_NewInt32(ctx, (int32_t)(val))
+#elif INT_MAX == INT64_MAX
+#define C_MACRO_INT_DEF(x) JS_PROP_INT64_DEF(#x, (int64_t)(x), JS_PROP_CONFIGURABLE)
+#define C_ENUM_DEF(x) JS_PROP_INT64_DEF(#x, (int64_t)(x), JS_PROP_CONFIGURABLE)
+#define JS_TO_INT(ctx, pres, val) JS_ToInt64(ctx, (int64_t *)(pres), val)
+#define JS_NEW_INT(ctx, val) JS_NewInt64(ctx, (int64_t)(val))
+#else
+#error "'int' neither 32bit nor 64 bit, I don't know how to handle it."
+#endif
+
+#define FFI_ALIGN(v, a)  (((((size_t) (v))-1) | ((a)-1))+1)
+
+#pragma region "FFI Helpers"
+// ===================
+
+static const char* ffi_strerror(ffi_status status){
+    switch(status){
```

Comment:
Can you run clang format? I think it will get rid of some of the style issues I am seeing, like missing spaces here.